Eventually pointed sequences (#1382)
UniMath/agda-unimath: src/foundation/eventually-constant-sequences.lagda.md

Reviewer: This file can be called `moduli-eventually-constant-sequences`, leaving room for the eventual file (no pun intended)`eventually-constant-sequences` which considers the proof theoretic correct notion of eventually constant sequences.

> # Eventually constant sequences

Reviewer: ```suggestion
# Moduli of eventually constant sequences
```

```agda
module foundation.eventually-constant-sequences where
```

<details><summary>Imports</summary>

```agda
open import elementary-number-theory.based-induction-natural-numbers
open import elementary-number-theory.inequality-natural-numbers
open import elementary-number-theory.maximum-natural-numbers
open import elementary-number-theory.natural-numbers

open import foundation.constant-maps
open import foundation.dependent-pair-types
open import foundation.eventually-equal-sequences
open import foundation.eventually-pointed-sequences-types
open import foundation.function-types
open import foundation.functoriality-dependent-pair-types
open import foundation.identity-types
open import foundation.sequences
open import foundation.universe-levels
```

</details>

## Idea

A [sequence](foundation.sequences.md) `u` is
{{#concept "eventually constant" Disambiguation="sequence" Agda=has-modulus-eventually-constant-sequence}}
if `u p ＝ u q` for sufficiently large `p` and `q`.

## Definitions

> ### Eventually constant sequences

Reviewer: ```suggestion
### Moduli of eventually constant sequences
```

```agda
module _
  {l : Level} {A : UU l} (u : sequence A)
  where

>   has-modulus-eventually-constant-sequence : UU l

Reviewer: Under your current definitions, you can't deduce that "an eventually constant sequence has a modulus" without a choice principle. By omitting the qualifier `has` you avoid this hairy problem with semantics. "Of course you can't extract a choice of a modulus from an eventually constant sequence without choice!"

```suggestion
  modulus-eventually-constant-sequence : UU l
```

This is analogous to how we say "countings of finite types" as opposed to "finite types have counting".

  has-modulus-eventually-constant-sequence =
    has-modulus-eventually-pointed-sequence
      (λ p → has-modulus-eventually-pointed-sequence (λ q → u p ＝ u q))
```

> ### The eventual value of an eventually constant sequence

Reviewer: I believe the standard nomenclature is to call this "eventual value" the "clustering point" or "asymptotic value". Please find a standard reference to follow for the terminology usage here.

```agda
module _
  {l : Level} {A : UU l} {u : sequence A}
  (H : has-modulus-eventually-constant-sequence u)
  where

  value-has-modulus-eventually-constant-sequence : A
  value-has-modulus-eventually-constant-sequence =
    u (modulus-has-modulus-eventually-pointed-sequence H)

  has-modulus-eventual-value-has-modulus-eventually-constant-sequence :
    has-modulus-eventually-pointed-sequence
      (λ n → value-has-modulus-eventually-constant-sequence ＝ u n)
  has-modulus-eventual-value-has-modulus-eventually-constant-sequence =
    value-at-modulus-has-modulus-eventually-pointed-sequence H
```

## Properties

### Constant sequences are eventually constant

```agda
module _
  {l : Level} {A : UU l} (x : A)
  where

  has-modulus-eventually-constant-const-sequence :
    has-modulus-eventually-constant-sequence (const ℕ x)
  pr1 has-modulus-eventually-constant-const-sequence = zero-ℕ
  pr2 has-modulus-eventually-constant-const-sequence p I =
    (zero-ℕ , λ _ _ → refl)
```

> ### An eventually constant sequence is eventually equal to the constant sequence of its eventual value

Reviewer: ```suggestion
### A sequence with a modulus of eventual constancy has a modulus of eventual equality to the constant sequence of its ??asymptotic value??
```
